src/v/bytes/details/io_fragment.h

```diff
@@ -94,14 +102,265 @@ class io_fragment {
+            _tail = &elem;
+        } else {
+            _head->_prev = &elem;
+        }
+        elem._next = _head;
+        _head = &elem;
+        update_generation();
+    }
+
+    io_fragment& back() {
+        check_consistency();
+        return *_tail;
+    }
+    const io_fragment& back() const {
+        check_consistency();
+        return *_tail;
+    }
+
+    void pop_back() {
+        check_consistency();
+        io_fragment* t = _tail;
+        _tail = _tail->_prev;
+        t->_prev = nullptr;
+        if (_tail != nullptr) {
+            _tail->_next = nullptr;
+        } else {
+            _head = nullptr;
+        }
+        update_generation();
+    }
+
+    void push_back(io_fragment& elem) {
+        assert(elem._prev == nullptr);
+        assert(elem._next == nullptr);
+        check_consistency();
+        if (_tail == nullptr) {
+            _head = &elem;
+        } else {
+            _tail->_next = &elem;
+        }
+        elem._prev = _tail;
+        _tail = &elem;
+        update_generation();
+    }
+
+    template<typename Func>
+    void clear_and_dispose(Func func) {
+        check_consistency();
+        io_fragment* current = _head;
+        while (current != nullptr) {
+            assert(current->_prev == nullptr);
+            io_fragment* next = current->_next;
+            if (next != nullptr) {
+                assert(next->_prev == current);
+                next->_prev = nullptr;
+            }
+            current->_next = nullptr;
+            func(current);
+            current = next;
+        }
+        _head = nullptr;
+        _tail = nullptr;
+        update_generation();
+    }
+
+    iterator begin() { return {this, _head}; }
+    iterator end() { return {this, nullptr}; }
+    reverse_iterator rbegin() { return {this, _tail}; }
+    reverse_iterator rend() { return {this, nullptr}; }
+    const_iterator cbegin() const { return {this, _head}; }
+    const_iterator cend() const { return {this, nullptr}; }
+
+private:
+    friend class iter<true, true>;
+    friend class iter<false, true>;
+    friend class iter<false, false>;
+    inline void update_generation() {
+#ifndef NDEBUG
+        ++_generation;
+#endif
+    }
+
+    inline void check_consistency() const {
+        // We don't check full consistency, just the head and tail elements, so
+        // we don't make debug mode *that* much slower.
+#ifndef NDEBUG
+        dassert(
+          (_head == nullptr) == (_tail != nullptr),
+          "head and tail consistency");
+        if (_head == nullptr) {
+            return;
+        }
+        dassert(!_head->_prev, "head must not have prev");
+        dassert(!_tail->_next, "tail must not have next");
+        dassert(
+          _head == _tail || _head->_next, "head must have next if size > 1");
+        dassert(
+          _head == _tail || _tail->_prev, "tail must have prev if size > 1");
+        dassert(
+          (_head->_next == _tail) != (_tail->_prev == _head),
```

Comment:
I think this should be == if i'm reading the assert message correctly

Reply:
Yes I also saw this and fixed :) thanks